bubble_sort.rb

```diff
@@ -0,0 +1,39 @@
+def bubble_sort(arr)
+  if arr.length == 1
+    return arr
+  end
+  swapped = true
+  while swapped
+    swapped = false
+    (0..arr.length-2).each do |idx|
+      if arr[idx] > arr[idx+1]
+        arr[idx], arr[idx+1] = arr[idx+1], arr[idx]
+        swapped = true
+      end 
+    end 
+  end
+  arr
+end 
+
+
+def bubble_sort_by(arr)
+  if arr.length == 1
+    return arr
+  end
+  swapped = true
+  while swapped
+    swapped = false
+    (0..arr.length-2).each do |idx|
+      comparison = yield arr[idx], arr[idx+1]
```

Comment:
So ... what this is doing is allowing you to sort however you like. In the "test" below, you've chosen to sort by the length of the string you're passing in. In general, with something like this a negative means left is "less than" right, 0 means equal, and positive means left is "greater than" right. You can use the technique here for interesting sorting methodologies. This is a good exercise, but the ruby enumerable (used by arrays etc.) you can do a similar thing.

Reply:
So ... had a few minutes this morning and came up with this ...
```ruby
janessa = OpenStruct.new(age: 24, height: 67)
jennifer = OpenStruct.new(age: 24, height: 60)
scott = OpenStruct.new(age: 29, height: 66)

sort_by_age = bubble_sort_by([janessa, jennifer, scott]) do |left, right|
  left.age - right.age
end
puts "sort_by_age = #{sort_by_age}"

sort_by_height = bubble_sort_by([janessa, jennifer, scott]) do |left, right|
  left.height - right.height
end
puts "sort_by_height = #{sort_by_height}"
```

which if you run in a console with your code you get ...

```
[19] pry(main)> janessa = OpenStruct.new(age: 24, height: 67)
=> #<OpenStruct age=24, height=67>
[20] pry(main)> jennifer = OpenStruct.new(age: 24, height: 60)
=> #<OpenStruct age=24, height=60>
[21] pry(main)> scott = OpenStruct.new(age: 29, height: 66)
=> #<OpenStruct age=29, height=66>
[22] pry(main)>
[23] pry(main)> sort_by_age = bubble_sort_by([janessa, jennifer, scott]) do |left, right|
[23] pry(main)*   left.age - right.age
[23] pry(main)* end
=> [#<OpenStruct age=24, height=67>, #<OpenStruct age=24, height=60>, #<OpenStruct age=29, height=66>]
[24] pry(main)> puts "sort_by_age = #{sort_by_age}"
sort_by_age = [#<OpenStruct age=24, height=67>, #<OpenStruct age=24, height=60>, #<OpenStruct age=29, height=66>]
=> nil
[25] pry(main)>
[26] pry(main)> sort_by_height = bubble_sort_by([janessa, jennifer, scott]) do |left, right|
[26] pry(main)*   left.height - right.height
[26] pry(main)* end
=> [#<OpenStruct age=24, height=60>, #<OpenStruct age=29, height=66>, #<OpenStruct age=24, height=67>]
[27] pry(main)> puts "sort_by_height = #{sort_by_height}"
sort_by_height = [#<OpenStruct age=24, height=60>, #<OpenStruct age=29, height=66>, #<OpenStruct age=24, height=67>]
=> nil
```

So, in the first case, we sort by the age of the people in the `OpenStruct` and in the 2nd case we sort by height.